### DISCUSSION TITLE
Is there a maximum number of credentials that can be issued per unit of time?

### DISCUSSION THREAD

**Discussion by JovanShandro** [decentralized-identity/veramo#638] [Q&A] [answered]
Hello!

I wanted to write a test that issued a large number of verifiable credentials to then test how much time it would take to get the credentials back from the data store. It was supposed to issue 100,000 credentials, however it only created 497 and no errors were thrown.

Is there a 'Maximum number of credentials that can be issued per second with the same issuer' or something similar?
Any idea what could be the reason for this?

If it can be of any help, I just used a for loop for creating the credentials:

```javascript
  for (let i = 0; i < 100 * 1000; i++) {
    const credential: W3CCredential = {
      issuer: { id: identifier.did }, // a did:ethr:rinkeby DID
      "@context": ["https://www.w3.org/2018/credentials/v1"],
      type: ["VerifiableCredential", "Profile"],
      issuanceDate: new Date().toISOString(),
      credentialSubject: {
        name: "Name",
        id: identifier.did,
      },
    };

    await agent.createVerifiableCredential({
      save: true,
      credential,
      proofFormat: "jwt",
    });
  }

```

Thank you!
Cheers

**Comment by mirceanis** [comment, marked as answer]
There is no limit imposed by Veramo.
I suspect that the reason you are only seeing a small number of credentials is because you are creating groups of identical credentials.
The only difference between the credentials you are creating is the timestamp, which has a resolution of Seconds.
When they are stored, they are hashed and the hash is used as primary key. since they are identical, the same hash is computed so they end up overwriting each other.

Please try to add a random nonce to your credentials and see if the issue manifests again.

```ts
import { v4 as uuidv4 } from 'uuid';
//...
  for (let i = 0; i < 100 * 1000; i++) {
    const credential: W3CCredential = {
      id: uuidv4(),
      issuer: { id: identifier.did }, // a did:ethr:rinkeby DID
      "@context": ["https://www.w3.org/2018/credentials/v1"],
      type: ["VerifiableCredential", "Profile"],
      issuanceDate: new Date().toISOString(),
      credentialSubject: {
        name: "Name",
        id: identifier.did,
      },
    };

    await agent.createVerifiableCredential({
      save: true,
      credential,
      proofFormat: "jwt",
    });
  }
```

**Reply by mirceanis** [reply]
Also, as a side-note, we plan to remove the `save` parameter from a bunch of agent methods, and rely instead on intentional data saving.
To create more future-proof code, you can try to write something like:
```ts
const verifiableCredential = await agent.createVerifiableCredential({
  save: false // or you can omit this parameter
  // ...
})
await agent.dataStoreSaveVerifiableCredential({ verifiableCredential })
```

**Reply by JovanShandro** [reply]
@mirceanis Thank you for the thorough explanation!
